java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -0,0 +1,643 @@
+      this.src = src;
+    }
+
+    private TaintProgramState(TaintProgramState other) {
+      this.src = other.src;
+      this.taintSources.putAll(other.taintSources);
+      this.methodCalls.addAll(other.methodCalls);
+    }
+
+    public TaintProgramState dup() {
+      return new TaintProgramState(this);
+    }
+
+    public SemanticModel semantic() {
+      return src.semantic();
+    }
+
+    private static String fullyQualify(Symbol s) {
+      if (s.isVariableSymbol()) {
+        Symbol owner = s.owner();
+        String ownerSignature;
+        if (owner.isTypeSymbol()) {
+          // field
+          ownerSignature = ((JavaSymbol.TypeJavaSymbol)owner).getFullyQualifiedName();
+        } else if (owner.isMethodSymbol()) {
+          // parameter
+          ownerSignature = ((JavaSymbol.MethodJavaSymbol)s.owner()).completeSignature();
+        } else {
+          throw new IllegalArgumentException();
+        }
+
+        return ownerSignature + "#" + s.name();
+      } else if (s.isMethodSymbol()) {
+        return ((JavaSymbol.MethodJavaSymbol)s).completeSignature();
+      } else {
+        throw new IllegalArgumentException();
+      }
+    }
+
+    public static boolean isString(Type type) {
+      return type.is("java.lang.String");
+    }
+
+    public TaintSource taintSourceFor(Symbol symbol) {
+      return taintSourceFor(symbol, Collections.emptyList());
+    }
+
+    public TaintSource taintSourceFor(Symbol symbol, List<TaintSource> arguments) {
+      if (symbol.isMethodSymbol()) {
+        MethodSymbol method = (MethodSymbol)symbol;
+
+        if (arguments.stream().anyMatch(ts -> ts.canBeTainted())) {
+          methodCalls.add(new DirectTaintSource(fullyQualify(symbol), arguments));
+        }
+
+        if (isString(method.returnType().type())) {
+          return new DirectTaintSource(fullyQualify(symbol), arguments);
+        } else {
+          return new TaintFreeSource();
+        }
+      }
+
+      if (symbol.isVariableSymbol() && symbol.owner().isTypeSymbol()) {
+        // Fields are not supported
+        return new TaintFreeSource();
+      }
+
+      return taintSources.computeIfAbsent(
+        symbol,
+        s -> new DirectTaintSource(fullyQualify(s), arguments));
+    }
+
+    public void put(Symbol symbol, TaintSource ts) {
+      taintSources.put(symbol, ts);
+    }
+
+    @Override
+    public int hashCode() {
+      return src.hashCode() +
+        13 * taintSources.hashCode() +
+        19 * methodCalls.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj.getClass().equals(this.getClass()))) {
+        return false;
+      }
+
+      TaintProgramState other = (TaintProgramState) obj;
+
+      return src.equals(other.src) &&
+        taintSources.equals(other.taintSources) &&
+        methodCalls.equals(other.methodCalls);
+    }
+
+  }
+
+  public static TaintSummary computeTaintConditions(ScannedFile src, CFG cfg) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)